usaspending_api/common/query_with_filters.py

```diff
@@ -0,0 +1,426 @@
+import logging
```

Comment:
I would recommend moving this or parts of this into package `usaspending_api/common/elasticsearch` since it has to do with Elasticsearch querying  and the module name by itself doesn't call that out.

If truly this is the case:
> The idea is that this ABC and QueryWithFilters can be altered to work with Django as well down the road.

Then following a kind of _strategy pattern_ (e.g. [here](https://refactoring.guru/design-patterns/strategy/python/example)), the abstract (backend agnostic) parts should be here in `usaspending_api/query_with_filters` then the elasticsearch-specific implementation would go in `usaspending_api/common/elasticsearch/query_with_filters.py` and django-specific implementation would go in `usaspending_api/common/django/query_with_filters.py`

Reply:
Worth further discussion though _**IF**_ we think there is value in being able to conjoin the "interface" to backend querying -- between Django and Elasticsearch. 

That is,  doing some queries or filters would use the same object as an abstraction/interface to a backend, and the developer wouldn't really need to know whether the backend was django or elasticsearch. 

My thought is that since Django is still going to be used heavily for various other endpoints (and likely in different ways than how it is/was used for Adv. Search), and Django will eventually be phased-out of Adv. Search ... that it is probably not worth the effort to make the two use a single backend querying interface. 

That shouldn't prevent you from good programming and good abstraction in building the Elasticsearch querying interface/layer -- just means that we can defer or choose not to build a django-specific implementation for it, which would likely require some compromises in the interface+impls to make it compatible for both.

Reply:
> and Django will eventually be phased-out of Adv. Search 

The plan is still to use Postgres for downloads. Advanced Search functionality contains downloads. As previously discussed there is a huge concern with creating dissonance between the future ES API results and download results. The idea of a shared filter object which can be reused for the various data stores and _hopefully_ ease the burden of keeping results consistent appears to be a clean design.

Reply:
But the downloads (from Adv Search) won't be running Django filters, will they? Isn't the approach to execute the filters on ES, get all the postgres IDs from the ES docs, and then run the download as a query to postgres with a giant kind of `IN` clause (no filtering aside from that)?

If you're thinking about the general Downloads section of the site (custom account/award download) we could adapt this interface to abstract those django queries. I'm not clear on how much similarity there is between the backend data structures of those (matviews?) and these ES indexes. If very close, then a common filtering interface could be good. But if they differ in what is filterable or the semantics of what some filterable field represents, forcing them to conform to the same interface could cause confusion (e.g. trying to add a filter that's not supported by one backed or another).

Reply:
You're correct on the current process of using the list of IDs from ES to drive downloads. While it is functional it plays into weaknesses of both Elasticsearch and Postgres, there is concern it will be the process long-term

Currently Custom Award Downloads use the same filters and SQL for downloads as Advanced Search.